cloudinit/sources/helpers/azure.py

```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
     return azure_endpoint
 
 
+@azure_ds_telemetry_reporter
+def http_with_retries(url, **kwargs) -> str:
+    exc = None
+
+    max_readurl_attempts = 240
+    default_readurl_timeout = 5
+    periodic_logging_attempts = 12
+
+    if 'timeout' not in kwargs:
+        kwargs['timeout'] = default_readurl_timeout
+
+    # remove kwargs that cause url_helper.readurl to retry,
+    # since we are already implementing our own retry logic.
+    if 'retries' in kwargs:
+        kwargs.pop('retries', None)
+        LOG.warning(
+            'retries kwarg passed in for communication with Azure endpoint.')
+    if 'infinite' in kwargs:
+        kwargs.pop('infinite', None)
+        LOG.warning(
+            'infinite kwarg passed in for communication with Azure endpoint.')
+
+    for attempt in range(max_readurl_attempts):
```

Comment:
We could change this call to range call to be 1-based instead of 0-based, then you don't have to add 1 to the attempt value below
```suggestion
    for attempt in range(1, max_readurl_attempts + 1):
```